### PR TITLE
Prometheus: Fix escaping of quotation marks for variables

### PR DESCRIPTION
**Pull request by gizmoguy**
**What is this feature?**

This PR fixes both regular and regex variable escaping in the prometheus data source for templated queries which use variables that contain double or single quotation marks.

This PR also adds escaping for adhoc filter label queries which allows adhoc filters to handle label values which contain backslashes or quotation marks.

**Why do we need this feature?**

There are a number of bugs in the current escaping behaviour when interpolating variables for prometheus, which can cause issues when querying for data in prometheus that use special characters in label values, such as `'`, `"` or `\`.

For example, if we have the following prometheus metrics:

```
singlequote_escape_test{label="looking'glass"} 1
doublequote_escape_test{label="looking\"glass"} 1
```

And create the following grafana dashboard variables:

```json
  "templating": {
    "list": [
      {
        "current": {
          "selected": true,
          "text": "looking'glass",
          "value": "looking'glass"
        },
        "hide": 0,
        "includeAll": false,
        "multi": false,
        "name": "singlequote_var",
        "options": [
          {
            "selected": true,
            "text": "looking'glass",
            "value": "looking'glass"
          }
        ],
        "query": "looking'glass",
        "queryValue": "",
        "skipUrlSync": false,
        "type": "custom"
      },
      {
        "current": {
          "selected": true,
          "text": "looking\"glass",
          "value": "looking\"glass"
        },
        "hide": 0,
        "includeAll": false,
        "multi": false,
        "name": "doublequote_var",
        "options": [
          {
            "selected": true,
            "text": "looking\"glass",
            "value": "looking\"glass"
          }
        ],
        "query": "looking\"glass",
        "queryValue": "",
        "skipUrlSync": false,
        "type": "custom"
      }
    ]
  },
```

If we try to use these variables in the following promql queries wrapped with double quotation marks (as is predominately the case for promql queries):

```
singlequote_escape_test{label="${singlequote_var}"}
doublequote_escape_test{label="${doublequote_var}"}
```

These queries will be interpolated as follows and sent to prometheus which will either not return the expected result, or trigger an error:

```
Expr: singlequote_escape_test{label="looking\\'glass"}    0 rows
Expr: doublequote_escape_test{label="looking"glass"}      Status: 400. Message: bad_data: invalid parameter "query": 1:40: parse error: unexpected identifier "glass" in label matching, expected "," or "}"
````

There is also a similar issue with adhoc filters where the label query made to prometheus when adding additional filters wasn't being escaped, so the query to `/api/v1/labels` has a similar issue:

```
request:    match[]	{label="looking"glass"}
response:   {"status":"error","errorType":"bad_data","error":"1:17: parse error: unexpected identifier \"glass\" in label matching, expected \",\" or \"}\""}
```

This PR addresses all of the above issues and allows dashboards using the prometheus datasource to properly handle variables that may contain special characters. The are also a number of test cases added to catch future regressions.

### CODE REVIEW

File: packages/grafana-prometheus/src/datasource.ts [hunk 1]
```diff
@@ -996,9 +996,14 @@
...
-  return typeof value === 'string' ? value.replace(/\\/g, '\\\\').replace(/'/g, "\\\\'") : value;
+  return typeof value === 'string' ? value.replace(/\\/g, '\\\\').replace(/"/g, '\\"') : value;
 }
 
 export function prometheusSpecialRegexEscape<T>(value: T) {
-  return typeof value === 'string' ? value.replace(/\\/g, '\\\\\\\\').replace(/[$^*{}\[\]\'+?.()|]/g, '\\\\$&') : value;
+  return typeof value === 'string'
+    ? value
+        .replace(/\\/g, '\\\\\\\\')
+        .replace(/"/g, '\\"')
```
[the remark this answers is not in the document]
Comment (gizmoguy) [reply packages/grafana-prometheus/src/datasource.ts]: @itsmylife Swapping lines 1005/1006 causes the query to have too many escaped backslashes which results in a syntax error back from prometheus:

```
bad_data: invalid parameter "query": 1:176: parse error: error parsing regexp: missing closing ): `(\\`
```

Instead I've added the appropriate number of backslashes now to prometheusSpecialRegexEscape() in e26b33d which should satisfy the codeql check.

**Review by NWRichmond — APPROVED** [review]
Thank you for this PR, @gizmoguy! I especially appreciate the detailed description.

While it'd be nice to limit the number of `.replace` calls in `prometheusSpecialRegexEscape`, I believe your solution represents the minimum number of `.replace` calls needed to achieve double escaping for backslashes, quotes, and other regex special chars.

@grafana/observability-metrics I've tested this out locally with good results. I'm not 100% sure that there won't be some unintended side-effects, but I believe it'd be healthiest to merge this change, monitor for regressions, and patch + add tests as needed.

**Review by itsmylife — APPROVED** [review]

**Comment by itsmylife** [comment]
Hi @gizmoguy, the tests work nicely. Thanks for the contribution. I'd like to merge this but there is a listing issue. Could you please fix it?

**Comment by gizmoguy** [comment]
@itsmylife Sorry about the linting issue, this has been fixed in the latest series of commits I just pushed.

**Comment by NWRichmond** [comment]
@gizmoguy unfortunately, we had to revert this PR after we discovered some experience-breaking side effects in [Explore Metrics](https://grafana.com/docs/grafana/latest/explore/simplified-exploration/metrics/).

I opened https://github.com/grafana/grafana/pull/96067, which I hope achieves the original goals of this PR. If you'd like to weigh in, please do!

**Comment by gizmoguy** [comment]
@NWRichmond apologies, I'll admit explore metrics isn't something I tested this PR against and this isn't a feature I use very often. I'm just spinning #96067 up in my dev environment and will add a review over there shortly.

**Comment by NWRichmond** [comment]
@gizmoguy no worries at all! Explore Metrics is still fairly new, and the regression didn't reveal itself until adding filters to a metrics exploration...tricky :)

Thank you for taking a look at #96067, we really appreciate it.

**Pull request merged**